presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -484,6 +484,20 @@ public PlanWithProperties visitSort(SortNode node, Context context)
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
+            else {
+                // current plan so far is single node, so local properties are effectively global properties
+                // skip the SortNode if the local properties guarantee ordering on Sort keys
+                // TODO: This should be extracted as a separate optimizer once the planner is able to reason about the ordering of each operator
+                List<LocalProperty<Symbol>> desiredProperties = new ArrayList<>();
+                for (Symbol symbol : node.getOrderBy()) {
+                    desiredProperties.add(new SortingProperty<>(symbol, node.getOrderings().get(symbol)));
+                }
+
+                if (LocalProperties.match(child.getProperties().getLocalProperties(), desiredProperties).stream()
+                        .noneMatch(Optional::isPresent)) {
+                    return child;
+                }
+            }
```

Comment:
I think it would be nice to separate `AddExchanges` into two optimizers ie. optimizer and visitor. Right now each `visit` method returns `PlanWithProperties`, in my approach `AddExchanges` would return `Plan` where `PhysicalPropertiesPlanVisitor` would return `Properties`.

Although, in this approach without any result memoization this could lead to `O(n^2)` cost of plan traversing. However, this could be simply solved somehow internally in the visitor.

Thanks to that `AddExchanges` could be separated to several optimizers. It also would help with rule migration process as there would more smaller optimizers to be ported, and properties visitor which would need to be mapped to [programmatic constraints](https://github.com/prestodb/presto/wiki/New-Optimizer#programmatic-constraints)

Do you such refactor is possible? Maybe I missed some other crucial part of `AddExchanges`... as it is typical with this class.